src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,642 @@
+                // `run_or_literal` and `run_of_zeroes_or_ones` slices include padding.
+                .run_or_literal = buf[offsets.run_or_literal..offsets.run_of_zeroes_or_ones],
+                .run_of_zeroes_or_ones = buf[offsets.run_of_zeroes_or_ones..offsets.run_lengths],
+                // `run_lengths`'s slice excludes the padding.
+                .run_lengths = buf[offsets.run_lengths..(offsets.run_lengths + self.runs)],
+                .literals = @alignCast(8,
+                    mem.bytesAsSlice(u64, buf[offsets.literals..offsets.end])),
+            };
+        }
+
+        fn encode(self: Self, dst: []u8, bitset: DynamicBitSetUnmanaged) void {
+            for (dst) |*b| b.* = 0;
+
+            const slices = self.parse(dst);
+            slices.total_blocks.* = @intCast(u32, bitset.bit_length);
+            slices.run_or_literal_count.* = @intCast(u32, self.runs + self.literals);
+            const word_count = bitsetMasksLen(bitset);
+
+            var run_or_literal_bit: usize = 0;
+            var run_index: usize = 0;
+            var literals_index: usize = 0;
+            var word_index: usize = 0;
+            while (word_index < word_count) {
+                const word = bitset.masks[word_index];
+                if (word == 0 or word == ~@as(usize, 0)) {
+                    const run_len = 1 + bitsetRunLength(bitset, word_index + 1, word_count, word);
+                    slices.run_lengths[run_index] = @intCast(u8, run_len - 1);
+                    if (word != 0) setBit(slices.run_of_zeroes_or_ones, run_index);
+                    setBit(slices.run_or_literal, run_or_literal_bit);
+                    run_index += 1;
+                    word_index += run_len;
+                } else {
+                    slices.literals[literals_index] = word;
+                    literals_index += 1;
+                    word_index += 1;
+                }
+                run_or_literal_bit += 1;
+            }
+
+            assert(word_index == word_count);
+            assert(slices.run_or_literal_count.* == run_or_literal_bit);
+            assert(slices.run_lengths.len == run_index);
+            assert(slices.literals.len == literals_index);
+        }
+
+        fn decode(data: []const u8, allocator: *mem.Allocator) !DynamicBitSetUnmanaged {
+            const run_or_literal_count = mem.readIntSliceLittle(u32, data[4..8]);
+            const run_or_literal_words = divCeil(usize, run_or_literal_count, 64);
+            const run_or_literal = mem.bytesAsSlice(u64, data[8 .. 8 + 8 * run_or_literal_words]);
+            var runs: usize = 0;
+            for (run_or_literal) |rol| runs += @popCount(u64, rol);
+            const decoder = BitsetEncoder(.immutable).init(.{
+                .runs = runs,
+                .literals = run_or_literal_count - runs,
+            });
+            const slices = decoder.parse(data);
+            assert(slices.run_or_literal_count.* == run_or_literal_count);
```

Comment:
> Is it possible to remove the parse() method and upfront @popCount() calculation and rather do the decoding (and encoding) all in a single pass?

I don't think that is possible with the current schema. See `BitsetEncoder.init` -- the offsets can only be computed if we know the precise number of runs and literals (independently, not the sum). 

But if we encode those counts separately rather than summed in `run_or_literal_count` then all of the offsets could be calculated after reading just the header.

Reply:
Yes, the latter was what I was thinking, so we have these counts stored at the front of the stream when we finish encoding. Then we can just decode in a single pass. Encoding should also be able to be a single pass, I believe?

Reply:
Or wait, can it? I think encoding still needs a first pass to count literals/runs, then a second pass to actually encode them. Until it has counted literals/runs, it doesn't know what offsets it should be encoding to.

Reply:
What if we tweak the output format a little?

We just work out the max possible size of both `run_of_zeroes_or_ones` and `run_lengths` (if all bits could be encoded as pure runs, as if there were no literals at all, leveraging our constant for max addressable bits) and then always have the encoding reserve this much space in the output. Then we know the offset of where the literals start. E.g. our metadata to record runs has fixed size in the layout.

Since runs are so efficient, reserving this little extra space is not a problem, it's very roughly on the order of a disk sector per 10 TiB I think?

Reply:
`run_or_literal` is variable-sized as well -- it is short when there are many long runs, but when there are many literals it is longer (it needs 1 bit for each literal or run). For example, 64 bits of `run_or_literal` could store `64 runs * 256 run words`, or it could store `64 literals * 1 literal per word`. That's an lot of padding in the former case to account for the latter in a fixed-size layout.

Reply:
Yes, the fixed size for `run_or_literal` assuming we have to support everything being a literal, would I think work out to 320 KiB (`10*1024*1024*1024*1024 / 65536 / 8 / 64 / 1024`) which is too much, dwarfing the superblock itself.

The cost of the extra pass over the data is also roughly 3ms if I understand correctly: `20,971,520 MiB / (6 GiB/s per-core bandwidth) * 1000ms` which means a further 15% throughput hit on the batch that follows a superblock checkpoint.

@sentientwaffle @ifreund This is an interesting curve ball, in that I didn't realize our scheme would need a second pass when encoding.

We could do away with the second pass if we encoded into a vector of buffers, and then did multiple I/O writes to write out these buffers (we don't do vectored I/O writes in TB to keep things simple) but that is too complex. Ideally, `encode()` should write into only a single large buffer and return how much was written.

Another way out is to shard the block free set when encoding, i.e. maybe leverage that earlier or later shards are all clean (all zeroes or all ones respectively, with literals only in the middle), but this is also adding complexity and getting too specific.

Our current scheme is blazing fast for decoding, but unfortunately not so great for encoding which is more important for our checkpointing use case.

Let's switch tack and simplify.

What about EWAH, also from Lemire & Co., an alternative to Roaring, a little slower but very simple and intuitive (even simpler then what we came up with) and with reasonable compression:

> EWAH may never (within 0.1%) generate a compressed bitmap larger than the uncompressed bitmap. It also uses only two types of words (see Fig. 1), where the first type is a 64-bit verbatim word. The second type of word is a marker word: the first bit indicates which clean word will follow, half the bits (32 bits) are used to store the number of clean words, and the rest of the bits (31 bits) are used to store the number of dirty words following the clean words. EWAH bitmaps begin with a marker word.

What do you think?

Reply:
EWAH sounds good to me. The extra-long runs (longer than 256 words) will be very useful when the blocks are mostly-allocated (or free). And the simplicity of the encoding/decoding is great.

Reply:
Let's do Option 3: i.e. `encode` returns the number of bytes encoded, and it's the caller's responsibility to store that. The caller would then slice the buffer so `decode` simply decodes everything it receives. In our case, we'll be storing the encoded size in the SuperBlock.

We shouldn't have a sentinel 'end-of-buffer' word at the end because that's less efficient (an extra branch in the hot loop compared to the caller just passing in the actual encoded slice to be decoded).